### ISSUE TITLE
Feature request: limited and unlimited randomized milestones and awards

### ISSUE THREAD

**Issue by vsrisuknimit** [bafolts/terraforming-mars#1311]
Some users voiced their opinions that they are fine with unlimited synergy milestones and awards.

Perhaps offer that as a choice.

(toggle) randomized milestones and awards 
  (radio box) limited synergy - [default]
  (radio box) no-limit synergy

For the limited synergy, I propose the following rules:
- no pair that is 7+ in synergy score (prohibit landlord & desert settler or terraformer & benefactor)
- not more than one pair that is 4+ in synergy score (allow one mayor & landlord or gardener & polar explorer)
- total less than or equal to 15 in synergy score

I think this is more flexible than the current max 1 synergy for all pairing rule.

**Comment by samyelias** [comment]
Are you referring to the synergy scores from https://ssimeonoff.github.io/ma-randomizer ? (Basically from https://boardgamegeek.com/thread/2023541/quantifying-milestone-and-awards-synergy-randomiza )

I forked my own version of ssimeonoff's randomizer, found here: https://samyelias.github.io/ma-randomizer . There are a few main differences: 

- I added a filter for max pair synergies; I personally never change the default value from 3 (excluding 4+ synergy pairs)
- I added weights to the randomization process to ensure a more equal probability of getting a particular goal. Otherwise, because of the synergy scores, certain goals were much more likely (like Tactician) and others much less likely (like Landlord
- Getting Desert Settler and Estate Dealer together is not good for any map except the one it first appeared on, so I made it randomly choose a map as well (shown in the readout).
- I tweaked the synergy scores compared to what's on those other sources: https://docs.google.com/spreadsheets/d/1dRpa9eUXjvPWg9-xyvaWrnD24I8ROXntFxRnswNBg40/edit#gid=330638583

Basically these are the excluded interactions I play with, that the weights for my randomizer are currently optimized for: 

TERRAFORMER and BENEFACTOR, 9
MAYOR and POLAR EXPLORER, 8
GARDENER and ECOLOGIST, 4
GARDENER and POLAR EXPLORER, 8
GARDENER and CULTIVATOR, 9
BUILDER and TYCOON, 4
BUILDER and CONTRACTOR, 9
SPECIALIST and ENERGIZER, 4
ECOLOGIST and DIVERSIFIER, 4
TYCOON and MAGNATE, 9
POLAR EXPLORER and LANDLORD, 4
POLAR EXPLORER and DESERT SETTLER, 8
ENERGIZER and INDUSTRIALIST, 6
HOVERLORD and EXCENTRIC, 5
LANDLORD and DESERT SETTLER, 7
LANDLORD and ESTATE DEALER, 7
LANDLORD and CULTIVATOR, 8
MINER and INDUSTRIALIST, 7
MINER and SPACE BARON, 4
MINER and CONTRACTOR, 4
INDUSTRIALIST and CONTRACTOR, 4
DESERT SETTLER and ESTATE DEALER, 5
DESERT SETTLER and CULTIVATOR, 7
ESTATE DEALER and CULTIVATOR, 8

I was thinking of implementing this in a fork of the web app if people were interested.

**Comment by kberg** [comment]
I like the care you put into a new analysis, samyelias. It's going to get
pretty hairy to have this kind of selection customized for individual
tastes. Consider for the web app merely giving full control over specific
selections, and have a separate app for weighing and selecting  MAs?

**Comment by samyelias** [comment]
I think a separate "manual" option for goals and milestones is a great idea either way.

As for including all the bells and whistles, I don't think it would be onerous. It would simply be a "balanced randomized goals" option. An additional option could be a slider for the synergy level as vsrisuknimit initially suggested, if people wanted that. The messy stuff is all under the hood.

**Comment by kberg** [comment]
I totally get it. But why is your algorithm balanced versus the other one?
Unless I'm missing something you see why I'd be concerned about a slippery
slope. (Not a very strong feeling, mind you, I'm just getting into the
discussion.)

 I do like the slider idea, tho.

>

**Comment by samyelias** [comment]
If we are simply comparing to totally random goals as the app now has, both ssimeonoff's and my version are better "balanced" because they try to avoid those high synergy goals to various extents. As a simple example of this, most people wouldn't want both the Industrialist award (Steel and Energy tokens) and the Miner award (Steel and Titanium tokens) both active in the same game, because of the obvious overlap.

If you mean compared to ssimeonoff's original randomizer page https://ssimeonoff.github.io/ma-randomizer : 

- I added weights to the randomization process to ensure a more equal probability of getting a particular goal (otherwise there is a bias towards the least synergistic goals)
- I added a filter for max pair synergies, to avoid any pair with too much synergy (as opposed to limiting the total synergy score, which I suppose if you kept it low enough would do the same thing)

So it is more "balanced" in that it allows for a more equal chance of getting all the different goals, and it is more "balanced" in a different way in that it doesn't allow for one big-synergy pair to sneak in under a "max total synergy" threshold.

Regarding the more equal chance and weights thing, this is not theoretical; we actually used the unweighted randomizer for a long time in my group, and then realized we were barely getting some goals while very often getting others. I ran the randomizer a million times and verified that the problem does in fact exist, so I added the weighting to fix it.
